domain/schema/model/sql/0022-resource.sql

```diff
@@ -0,0 +1,108 @@
+CREATE TABLE resource_origin_kind (
+    id INT PRIMARY KEY,
+    name TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_resource_origin_name
+ON resource_origin_kind (name);
+
+INSERT INTO resource_origin_kind VALUES
+(0, 'uploaded'),
+(1, 'store');
+
+CREATE TABLE resource_state (
+    id INT PRIMARY KEY,
+    name TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_resource_state
+ON resource_state (name);
+
+INSERT INTO resource_state VALUES
+(0, 'available'),
+(1, 'potential');
+
+CREATE TABLE resource (
+    uuid TEXT NOT NULL PRIMARY KEY,
+    application_uuid TEXT NOT NULL,
+    name TEXT NOT NULL,
+    origin_kind_id INT NOT NULL,
+    state_id INT NOT NULL,
+    size INT,
+    fingerprint TEXT,
+    created_at TIMESTAMP NOT NULL,
+    CONSTRAINT fk_resource_application_uuid
+    FOREIGN KEY (application_uuid)
+    REFERENCES application (uuid),
+    CONSTRAINT fk_resource_name
+    FOREIGN KEY (name)
+    REFERENCES resource_meta (name),
+    CONSTRAINT fk_resource_origin_kind_id
+    FOREIGN KEY (origin_kind_id)
+    REFERENCES resource_origin_kind (id),
+    CONSTRAINT fk_resource_state_id
+    FOREIGN KEY (state_id)
+    REFERENCES resource_state (id)
+);
+
+CREATE UNIQUE INDEX idx_resource ON resource (application_uuid, name, state_id);
+
+CREATE TABLE resource_type (
+    id INT PRIMARY KEY,
+    name TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_resource_type
+ON resource_type (name);
+
+INSERT INTO resource_type VALUES
+(0, 'file'),
+(1, 'oci-image');
+
+CREATE TABLE resource_meta (
+    application_uuid TEXT NOT NULL,
+    name TEXT NOT NULL,
+    type_id INT NOT NULL,
+    path TEXT NOT NULL,
+    description TEXT,
+    CONSTRAINT fk_resource_type_id
+    FOREIGN KEY (type_id)
+    REFERENCES resource_type (id),
+    PRIMARY KEY (application_uuid, name)
+);
+
+CREATE UNIQUE INDEX idx_resource_meta ON resource_meta (application_uuid, name);
+
+CREATE TABLE application_resource (
+    resource_uuid TEXT NOT NULL PRIMARY KEY,
+    -- Empty by default, user id of who uploaded the resource, or unit id
+    -- of the unit which triggered download from a repository.
+    supplied_by TEXT,
```

Comment:
I'll be honest, I don't like this. If we don't want to keep RI, because the supplier has been removed either the unit or the user, then we should at least have a `supplier_type_id` which is an enum of `user` or `unit`. Then `supplied_by` should then become `supplied_by_uuid`.

Reply:
We don't need to keep RI here - This data is used in juju cli output, either as the name of a user, or as the word `store` as a generic word for charm hub.

```
$ juju resources application
Resource  Supplied by  Revision
foo-file  store        3
```

We might just need the name of the user, rather than a uuid, as if the user is deleted I don't believe that their deployed applications disappear. Let me investigate what actually happens.

Reply:
Even after the user heather has been removed.
```
$ juju resources juju-qa-test
Resource  Supplied by  Revision
foo-file  heather      2024-08-01T16:10
```

Reply:
done and added a third kind of supplied_by: application which is used for oci-images.

Reply:
Depending on the data, you have the right to be forgotten. So yeah this is kind of interesting.